src/Swarm/Language/LSP/Hover.hs

```diff
@@ -130,23 +163,67 @@ explain = \case
   TRef {} -> pure $ pureDoc "A memory reference.  These likewise never show up in surface syntax but are here to facilitate pretty-printing."
   TRequireDevice {} -> pure $ pureDoc "Require a specific device to be installed."
   TRequire {} -> pure $ pureDoc "Require a certain number of an entity."
-  TVar x -> pure $ pureDoc $ "var: " <> U.bquote x
-  SLam {} -> pure $ pureDoc "A lambda expression, with or without a type annotation on the binder."
-  SApp (Syntax _ t1) (Syntax _ t2) ->
-    Node
-      (pureDoc "Function application of:")
-      [explain t1, explain t2]
-  SLet isRecursive _ maybeTypeAnnotation _ _ ->
-    pure $
-      pureDoc $
-        T.unwords
-          [ "A"
-          , (if isRecursive then "" else "non-") <> "recursive"
-          , "let expression"
-          , if null maybeTypeAnnotation then "without" else "with"
-          , "a type annotation on the variable."
-          ]
+  TVar v -> pure $ pureDoc $ "`" <> v <> ": " <> maybe "?" prettyText (Ctx.lookup v . moduleCtx =<< mMod) <> "`"
+  SLam v _mType _syn -> pure . pureDoc $ "A lambda expression binding the variable " <> U.bquote v <> "."
+  SApp _ _ -> explainFunction len mMod trm
+  SLet isRecursive var maybeTypeAnnotation _r _b -> pure $ explainDefinition len False mMod isRecursive var maybeTypeAnnotation
+  SDef isRecursive var maybeTypeAnnotation _body -> pure $ explainDefinition len True mMod isRecursive var maybeTypeAnnotation
   SPair {} -> pure $ pureDoc "A pair."
-  SDef {} -> pure $ pureDoc "A (recursive) definition command, which binds a variable to a value in subsequent commands."
   SBind {} -> pure $ pureDoc "A monadic bind for commands, of the form `c1 ; c2` or `x <- c1; c2`."
   SDelay {} -> pure $ pureDoc "Delay evaluation of a term, written `{...}`.  Swarm is an eager language, but in some cases (e.g. for `if` statements and recursive bindings) we need to delay evaluation.  The counterpart to `{...}` is `force`, where `force {t} = t`. Note that 'Force' is just a constant, whereas 'SDelay' has to be a special syntactic form so its argument can get special treatment during evaluation."
+
+-- | Helper function to explain function application.
+--
+-- Note that 'Force' is often inserted internally, so
+-- if it shows up here we drop it.
+--
+-- >>> tprint = error . ("This is fine:\n" <>) . T.unpack
+-- >>> dprint = tprint . treeToMarkdown 0
+-- >>> dprint $ explainFunction Long Nothing (mkOp' Mul (TInt 1) (TInt 2)) -- 1 * 2
+-- *** Exception: This is fine:
+-- Function application of:
+-- * `*: int -> int -> int`
+--   Multiply the given integer values.
+-- * with parameters:
+--     * An integer literal.
+--     * An integer literal.
+explainFunction :: Length -> Maybe TModule -> Term -> Tree (DocLine Text)
+explainFunction len mMod t =
+  case unfoldApps t of
+    (TConst Force :| [innerT]) -> explain len mMod innerT
+    (TConst Force :| f : params) -> explainF f params
```

Comment:
This is a fascinating problem - the internally inserted `Force` is shown instead of the inner terms.

If we kept the locations, this workaround might be unnecessary.

![image](https://user-images.githubusercontent.com/44544735/211225077-e0587dba-6e9c-49ed-a033-93a076a6008e.png)

Reply:
I suspect force uses pattern matching that gets rid of the source position.

My hope was that the inferred types would be stored in the AST (def/let/…) but that is not the case.

@byorgey is there any useful (in this context) change in the syntax, or should I just use the original?